DECISION-RESEARCH.md

```diff
@@ -0,0 +1,315 @@
+
+---
+
+## Updating resources
+
+**Decision: Use PATCH and json-patch **
+
+### Patch
+
+When deciding on the update strategy of resources in our API we took into consideration two different proposals for sending updates: Payload comparison
+
+The key reasons were:
+
+- Wikidata works with large JSON objects
+- json-patch Allows appending to arrays
+
+Payload comparison: https://www.notion.so/Payload-Examples-1cbe38a24f8044fba18d107d390b97fb
+
+#### RFC 6902 (json-patch+json)
+
+- More php tooling and support (not extremely popular or regularly updated but it is an old RFC):
+  - github.com/mikemccabe/json-patch-php
+  - github.com/swaggest/json-diff 
+- Explicit approach with predefined available actions
+- Allows appending to arrays
+- Multiple operations are performed by sending an array of “operation” objects
+
+#### RFC 7396 (merge-patch+json)
+
+- Straightforward, minimal verbosity
+- Removal is “implicit”
+  - i.e. by sending a key with a `null` value
+- No support for extending array values 
+  - i.e. arrays must be fully replaced
+- PHP tooling seems to be scarce and mostly abandoned
+- Multiple operations are performed by sending one “update” object
+
+---
+
+## Entities + Statements
+
+**Decision: Statements would be exposed both within an entity and also at the top level**
+
+Considered options:
+
+- For the statements with a GUID:
+  - Should we address those always through entities?
+    - longer path
+    - the entity id is already in the path and the prefix is redundant
+- Or individually, through a /statements path? 
+
+Usefulness (end users) Vs. cost of ownership (devs)
+
+### Notes
+
+The current Action API allows you to get a statement with only the GUID (without the entity id prefix) with no extra work
+
+Not having a top level statement endpoint would mean that “extra work” would be needed by the client in order to access a statement (splitting the statement ID)
+
+Taking this into account we can cover all of our existing APIs with REST APIs only if we do have statements as a top level api module.
+
+---
+
+## RevisionIds (Path? Query?)
+
+**Decision: We will not worry about revisionids in paths, as the current API doesn't expose them and per core behaviour this should probably be left for the revisions endpoint**
+
+Both choices seem to be okay according to stackoverflow https://stackoverflow.com/a/12737540/4746236
+
+Mediawiki core REST API only has revisionid when looking specifically at revisions (not pages) https://www.mediawiki.org/wiki/API:REST_API
+
+- In the query:
+  - Drupal https://www.drupal.org/docs/core-modules-and-themes/core-modules/jsonapi-module/revisions
+- In the Path:
+  - Wikimedia (RestBase) https://en.wikipedia.org/api/rest_v1/#/Page%20content/getFormatRevision
+  - Wordpress https://developer.wordpress.org/rest-api/reference/page-revisions/#retrieve-a-page-revision
+
+---
+
+## Conditional Requests
+
+**Decision: We will implement conditional requests for gets.**
```

Comment:
I'm reading this as 

> implement conditional requests for GETs

and we [see this implemented for them](https://github.com/wmde/wikibase-rest-api-proposal/blob/6499299dead81c862c4ccf69af237d6fd579c65d/specs/resources/items/single.json#L9).

The notes however refer to how _update_ problems can be prevented - i.e. helping the server make a call if and how a change can be applied (e.g. by informing a merge of a "baseRevId") as illustrated by the PUT example in https://www.novatec-gmbh.de/en/blog/managing-concurrency-in-a-distributed-restful-environment-with-spring-boot-and-angular2/

Reply:
The problem with conditional requests for requests that edit that I see are that If-None-Match and If-Match / the conditional request model do not really work for our current expectations.

When you make an edit and you supply the baserevid of 1, but the current revision is actually 2, you do not want the request to fail, you want the application to decide if it can perform the merge, and only fail if that is not possible, which I see as the reason that the currently implemented mediawiki core API only does this for reads and sticks to the more flexible latest id and or timestamp in the payload.

Having said that, I guess there is no reason we couldn't also implement conditional requests for writes? as well as the latest data in the payload?

Potentially we should be thinking about this merging pattern slightly differently for some of our APIs, such as PATCH? but I figure that is another rabbit hole.

Reply:
> When you make an edit and you supply the baserevid of 1, but the current revision is actually 2, you do not want the request to fail,

That is understood. But while you are describing a behavior like e.g. [here](https://docs.microsoft.com/en-us/rest/api/storageservices/merge-entity#remarks) I expected (maybe incorrectly?) it to be up to the server to use the information from the header to try for a merge and only respond with 409 if it can't find a solution. I'm fairly certain I did not dream up the idea but can't find the/a link at the moment.

Reply:
Indeed perhaps we could have less strict behaviour. Simply reading `If-Match` makes it sounds very strict but diving back into the docs `If the request cannot be fulfilled, the 412 (Precondition Failed) response is returned.`.
`cannot be fulfilled` !== the match was not made.

In that case it would be an interesting line of questioning once again for the core platform team and the decision that they made there (which is ultimately what we are copying)

Reply:
I have expanded the decision